README.md
# Custom Map Markers

Not only "*Everything is a widget*" also "*Everywhere is a widget*".  
`custom_map_marker` is a package that turns runtime widgets into map marker icons.

<img src="https://github.com/IbrahimTabba/custom_map_markers/blob/master/example/images/custom-map-markers.png?raw=true" width="256">

## Features

this package providers the following features:

- Ability to create runtime map markers using flutter widgets.
- Capture Images of any list of widgets.


## Getting started

First, add `custom_map_markers` as a dependency in your pubspec.yaml file.
```yaml
dependencies:
  custom_map_markers: ^0.0.1
```


## Usage


1. Wrap your `GoogleMaps` widget with `CustomGoogleMapMarkerBuilder`.
2. Define each marker and its custom widget in `customMarkers`
3. Once custom markers are ready they will be passed as `Set<Marer>` in `builder` function.


```dart  
final locations = const [  
  LatLng(37.42796133580664, -122.085749655962),  
  LatLng(37.41796133580664, -122.085749655962),  
];  

CustomGoogleMapMarkerBuilder(  
  customMarkers: [  
  MarkerData(  
        marker: Marker(  markerId: const MarkerId('id-1'), position: locations[0]),  
		child: _customMarker('A', Colors.black)),  
  MarkerData(  
        marker: Marker(  markerId: const MarkerId('id-2'), position: locations[1]),  
		child: _customMarker('B', Colors.red)),    
  ],  
  builder: (BuildContext context, Set<Marker>? markers) {  
    if (markers == null) {  
      return const Center(child: CircularProgressIndicator());  
  }  
    return GoogleMap(  
      initialCameraPosition: const CameraPosition(  
        target: LatLng(37.42796133580664, -122.085749655962),  
  zoom: 14.4746,  
  ),  
  markers: markers,  
  onMapCreated: (GoogleMapController controller) { },  
  );  
  },  
)
```  

**Not a google map user?**

> if you are not willing to use google maps you can use
> `CustomMapMarkerBuilder` that takes `List<widgets>` instead  of `List<MarkerData>` and returns `List<Uint8List>` instead of `Set<Marker>` so you are free to use captured png images as you need.

### Notes:

1. Make sure your marker widget has boundaries.
2. This package capture a png image of widget markers once they render, if your widget would take more time to render like `Image.network` you can use `screenshotDelay` parameter to daley the capture process making sure that your widget is rendered properly before capturing it.
3. to run project example you have to add a valid google maps key to android/ios directories.


## Features and bugs
Please feel free to:

-   file feature requests and bugs at the  [issue tracker](https://github.com//IbrahimTabba/custom_map_markers/issues)
  

CHANGELOG.md
## 0.0.1

* initial release

## 0.0.2

* Fix marker is always null in release mode

## 0.0.2+1

* Support updating custom markers at runtime


lib/src/custom_marker.dart
import 'dart:typed_data';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:collection/collection.dart';
import 'dart:ui' as ui;
import 'package:synchronized/synchronized.dart';

/// a widgets that capture a png screenshot of its content and pass it through
/// [onImageCaptured].
class CustomMarker extends StatefulWidget {
  final Widget child;
  final Function(Uint8List?)? onImageCaptured;
  final Duration? screenshotDelay;

  const CustomMarker(
      {Key? key,
      required this.child,
      this.onImageCaptured,
      this.screenshotDelay})
      : super(key: key);

  @override
  _CustomMarkerState createState() => _CustomMarkerState();
}

class _CustomMarkerState extends State<CustomMarker> {
  final GlobalKey key = GlobalKey();
  final Function eq = const ListEquality().equals;
  Uint8List? _lastImage;
  final lock = Lock();

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    WidgetsBinding.instance?.addPostFrameCallback((_) {
      lock.synchronized(() async {
        await Future.delayed(
            widget.screenshotDelay ?? const Duration(milliseconds: 500));
        final _image = await _capturePng(key);
        if (_lastImage == null || !eq(_lastImage!, _image)) {
          _lastImage = _image;
          widget.onImageCaptured?.call(_image);
        } else {
          widget.onImageCaptured?.call(_lastImage);
        }
      });
    });
    return RepaintBoundary(
      key: key,
      child: widget.child,
    );
  }

  Future<Uint8List?> _capturePng(GlobalKey iconKey) async {
    try {
      final RenderRepaintBoundary? boundary =
          iconKey.currentContext?.findRenderObject() as RenderRepaintBoundary?;
      if (kDebugMode && (boundary?.debugNeedsPaint ?? false)) {
        await Future.delayed(const Duration(milliseconds: 200));
        return _capturePng(iconKey);
      }
      ui.Image? image = await boundary?.toImage(pixelRatio: 3.0);
      ByteData? byteData =
          await image?.toByteData(format: ui.ImageByteFormat.png);
      var pngBytes = byteData?.buffer.asUint8List();
      return pngBytes;
    } catch (e) {
      return null;
    }
  }
}


lib/src/custom_map_marker_builder.dart
import 'dart:typed_data';

import 'package:custom_map_markers/src/custom_marker.dart';
import 'package:flutter/material.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

/// A widget that converts a list of widget into png images
/// returned as [Uint8List].

class CustomMapMarkerBuilder extends StatelessWidget {
  /// List of custom widgets that are going to be used as map markers.
  final List<Widget> markerWidgets;

  /// Each widget in [markerWidgets] will be converted into png image inorder
  /// be used as map marker.
  /// [screenshotDelay] will add an extra delay before taking screenshot
  /// of widgets.
  ///
  /// You can use this if you widget needs more time to build or render
  /// like [Image.network] for example.
  final Duration screenshotDelay;

  /// Widget builder that carries [imagesData] which is `null` when marker
  /// images are not ready yet or list of [Uint8List] when custom markers
  /// are ready.
  final Widget Function(BuildContext context, List<Uint8List>? imagesData)
      builder;

  /// [controller] controls the state of captured images from widgets.
  late final _MarkersController controller;

  CustomMapMarkerBuilder({
    Key? key,
    required this.markerWidgets,
    required this.builder,
    this.screenshotDelay = const Duration(milliseconds: 500),
  }) : super(key: key) {
    controller = _MarkersController(
        value: List<Uint8List?>.filled(markerWidgets.length, null),
        childCount: markerWidgets.length);
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        ...List.generate(
            markerWidgets.length,
            (index) => Positioned(
                  left: -MediaQuery.of(context).size.width,
                  child: CustomMarker(
                    key: ObjectKey(markerWidgets[index]),
                    child: markerWidgets[index],
                    screenshotDelay: screenshotDelay,
                    onImageCaptured: (data) {
                      controller.updateRenderedImage(index, data);
                    },
                  ),
                )),
        ValueListenableBuilder(
            valueListenable: controller,
            builder:
                (BuildContext context, List<Uint8List?> value, Widget? child) {
              return builder(context, controller.images);
            })
      ],
    );
  }
}

/// A widget convert list of widgets into custom google maps marker icons.

class CustomGoogleMapMarkerBuilder extends StatelessWidget {
  /// List of custom [MarkerData] each item is a google maps [Marker] and
  /// custom widget thar are going to be used as an icon for this marker.
  final List<MarkerData> customMarkers;

  /// Each widget in [markerWidgets] will be converted into png image inorder
  /// be used as map marker.
  /// [screenshotDelay] will add an extra delay before taking screenshot
  /// of widgets.
  ///
  /// You can use this if you widget needs more time to build or render
  /// like [Image.network] for example.
  final Duration screenshotDelay;

  /// Widget builder that carries [imagesData] which is `null` when marker
  /// images are not ready yet or list of google maps [Marker] when custom
  /// markers are ready.
  final Widget Function(BuildContext, Set<Marker>? markers) builder;

  const CustomGoogleMapMarkerBuilder(
      {Key? key,
      required this.customMarkers,
      required this.builder,
      this.screenshotDelay = const Duration(milliseconds: 500)})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return CustomMapMarkerBuilder(
      screenshotDelay: screenshotDelay,
      markerWidgets:
          customMarkers.map((customMarker) => customMarker.child).toList(),
      builder: (BuildContext context, List<Uint8List>? customMarkerImagesData) {
        return builder(
            context,
            customMarkerImagesData == null
                ? null
                : customMarkers
                    .map((e) => e.marker.copyWith(
                        iconParam: BitmapDescriptor.fromBytes(
                            customMarkerImagesData[customMarkers.indexOf(e)])))
                    .toSet());
      },
    );
  }
}

/// [MarkerData] carries google maps marker and its desired icon widget
class MarkerData {
  final Marker marker;
  final Widget child;

  MarkerData({required this.marker, required this.child});
}

/// [_MarkersController] handles the state of rendered markers and notify
/// listeners when all marker are rendered and captured.

class _MarkersController extends ValueNotifier<List<Uint8List?>> {
  final int childCount;

  late final List<Uint8List?> renderedWidgets;

  _MarkersController(
      {required List<Uint8List?> value, required this.childCount})
      : super(value) {
    renderedWidgets = List<Uint8List?>.filled(childCount, null);
  }

  updateRenderedImage(int index, Uint8List? data) {
    renderedWidgets[index] = data;
    if (ready) {
      value = List.from(renderedWidgets);
    }
  }

  bool get ready => !renderedWidgets.any((image) => image == null);

  List<Uint8List>? get images => ready ? value.cast<Uint8List>() : null;
}


example/lib/main.dart
import 'package:custom_map_markers/custom_map_markers.dart';
import 'package:flutter/material.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Custom Marker Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Custom Marker Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final locations = const [
    LatLng(37.42796133580664, -122.085749655962),
    LatLng(37.41796133580664, -122.085749655962),
    LatLng(37.43796133580664, -122.085749655962),
    LatLng(37.42796133580664, -122.095749655962),
    LatLng(37.42796133580664, -122.075749655962),
  ];

  late List<MarkerData> _customMarkers;

  @override
  void initState() {
    super.initState();
    _customMarkers = [
      MarkerData(
          marker:
              Marker(markerId: const MarkerId('id-1'), position: locations[0]),
          child: _customMarker3('Everywhere\nis a Widgets', Colors.blue)),
      MarkerData(
          marker:
              Marker(markerId: const MarkerId('id-5'), position: locations[4]),
          child: _customMarker('A', Colors.black)),
      MarkerData(
          marker:
              Marker(markerId: const MarkerId('id-2'), position: locations[1]),
          child: _customMarker('B', Colors.red)),
      MarkerData(
          marker:
              Marker(markerId: const MarkerId('id-3'), position: locations[2]),
          child: _customMarker('C', Colors.green)),
      MarkerData(
          marker:
              Marker(markerId: const MarkerId('id-4'), position: locations[3]),
          child: _customMarker2('D', Colors.purple)),
      MarkerData(
          marker:
              Marker(markerId: const MarkerId('id-5'), position: locations[4]),
          child: _customMarker('A', Colors.blue)),
    ];
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          setState(() {
            if (_customMarkers.isNotEmpty) {
              _customMarkers.removeLast();
            }
          });
        },
      ),
      body: CustomGoogleMapMarkerBuilder(
        //screenshotDelay: const Duration(seconds: 4),
        customMarkers: _customMarkers,
        builder: (BuildContext context, Set<Marker>? markers) {
          if (markers == null) {
            return const Center(child: CircularProgressIndicator());
          }
          return GoogleMap(
            initialCameraPosition: const CameraPosition(
              target: LatLng(37.42796133580664, -122.085749655962),
              zoom: 14.4746,
            ),
            markers: markers,
            onMapCreated: (GoogleMapController controller) {},
          );
        },
      ),
    );
  }

  _customMarker(String symbol, Color color) {
    return Stack(
      children: [
        Icon(
          Icons.add_location,
          color: color,
          size: 50,
        ),
        Positioned(
          left: 15,
          top: 8,
          child: Container(
            width: 20,
            height: 20,
            decoration: BoxDecoration(
                color: Colors.white, borderRadius: BorderRadius.circular(10)),
            child: Center(child: Text(symbol)),
          ),
        )
      ],
    );
  }

  _customMarker2(String symbol, Color color) {
    return Container(
      width: 30,
      height: 30,
      margin: const EdgeInsets.all(8),
      decoration: BoxDecoration(
          border: Border.all(color: color, width: 2),
          color: Colors.white,
          borderRadius: BorderRadius.circular(15),
          boxShadow: [BoxShadow(color: color, blurRadius: 6)]),
      child: Center(child: Text(symbol)),
    );
  }

  _customMarker3(String text, Color color) {
    return Container(
      margin: const EdgeInsets.all(8),
      padding: const EdgeInsets.all(8),
      decoration: BoxDecoration(
          border: Border.all(color: color, width: 2),
          color: Colors.white,
          borderRadius: BorderRadius.circular(4),
          boxShadow: [BoxShadow(color: color, blurRadius: 6)]),
      child: Center(
          child: Text(
        text,
        textAlign: TextAlign.center,
      )),
    );
  }
}
